clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -3990,9 +3991,16 @@ QualType Sema::CheckTemplateIdType(TemplateName Name,
     if (Inst.isInvalid())
       return QualType();
 
-    CanonType = SubstType(Pattern->getUnderlyingType(),
-                          TemplateArgLists, AliasTemplate->getLocation(),
-                          AliasTemplate->getDeclName());
+    if (!RebuildingTypesInCurrentInstantiation) {
```

Comment:
This is probably better/less likely to be screwed up in the future if we instead immediately do a 'SavedContext.pop()` if this condition is true, which leaves the SubstType code from being repeated/accidentially diverging.